File: sdk/typescript/src/retry.ts
export interface RestRetryOptions {
  maxAttempts?: number;
  initialBackoffMs?: number;
  maxBackoffMs?: number;
  jitter?: number;
  onRetry?: (event: RestRetryEvent) => void;
}

export interface RestRetryEvent {
  attempt: number;
  delayMs: number;
  error?: unknown;
  response?: Response;
}

type SDKResult = { response?: Response };

/** Retries a generated REST usage call with server-directed backpressure. */
export async function retryUsageWrite<T extends SDKResult>(call: () => Promise<T>, options: RestRetryOptions = {}): Promise<T> {
  const policy = normalize(options);
  for (let attempt = 1; ; attempt++) {
    let result: T | undefined;
    let error: unknown;
    try {
      result = await call();
      if (!retryableStatus(result.response?.status)) {
        return result;
      }
    } catch (caught) {
      if (!(caught instanceof TypeError)) {
        throw caught;
      }
      error = caught;
    }
    if (attempt >= policy.maxAttempts) {
      if (error !== undefined) throw error;
      return result!;
    }
    const delayMs = retryDelay(result?.response, policy, attempt);
    policy.onRetry?.({ attempt: attempt + 1, delayMs, error, response: result?.response });
    await new Promise((resolve) => setTimeout(resolve, delayMs));
  }
}

function normalize(options: RestRetryOptions) {
  const initialBackoffMs = Math.max(1, options.initialBackoffMs ?? 100);
  return {
    maxAttempts: Math.max(1, Math.floor(options.maxAttempts ?? 3)),
    initialBackoffMs,
    maxBackoffMs: Math.max(initialBackoffMs, options.maxBackoffMs ?? 5_000),
    jitter: Math.min(1, Math.max(0, options.jitter ?? 0.2)),
    onRetry: options.onRetry,
  };
}

function retryableStatus(status: number | undefined): boolean {
  return status === 429 || status === 500 || status === 502 || status === 503 || status === 504;
}

function retryDelay(response: Response | undefined, policy: ReturnType<typeof normalize>, attempt: number): number {
  const retryAfter = response?.headers.get("Retry-After")?.trim();
  if (retryAfter) {
    const seconds = Number(retryAfter);
    if (Number.isFinite(seconds) && seconds > 0) return Math.ceil(seconds * 1_000);
    const date = Date.parse(retryAfter);
    if (Number.isFinite(date) && date > Date.now()) return date - Date.now();
  }
  const backoff = Math.min(policy.maxBackoffMs, policy.initialBackoffMs * 2 ** (attempt - 1));
  return Math.max(1, Math.round(backoff * (1 - policy.jitter + Math.random() * 2 * policy.jitter)));
}

File: sdk/python/open_spanner_client/retry.py
"""Opt-in retries for generated REST usage writes."""

from __future__ import annotations

import asyncio
import random
import time
from collections.abc import Awaitable, Callable
from dataclasses import dataclass
from datetime import UTC, datetime
from email.utils import parsedate_to_datetime
from typing import TypeVar

import httpx

from open_spanner_client.types import Response


@dataclass(frozen=True)
class RestRetryEvent:
    attempt: int
    delay_seconds: float
    error: httpx.TransportError | None = None
    response: Response[object] | None = None


@dataclass(frozen=True)
class RestRetryPolicy:
    max_attempts: int = 3
    initial_backoff_seconds: float = 0.1
    max_backoff_seconds: float = 5.0
    jitter: float = 0.2
    on_retry: Callable[[RestRetryEvent], None] | None = None


T = TypeVar("T")


def retry_usage_write(call: Callable[[], Response[T]], policy: RestRetryPolicy | None = None) -> Response[T]:
    """Retry a generated ``sync_detailed`` usage call."""
    normalized = _normalize(policy or RestRetryPolicy())
    for attempt in range(1, normalized.max_attempts + 1):
        response: Response[T] | None = None
        error: httpx.TransportError | None = None
        try:
            response = call()
            if not _retryable_status(response.status_code):
                return response
        except httpx.TransportError as caught:
            error = caught
        if attempt >= normalized.max_attempts:
            if error is not None:
                raise error
            return response  # type: ignore[return-value]
        delay = _delay(response, normalized, attempt)
        if normalized.on_retry is not None:
            normalized.on_retry(RestRetryEvent(attempt + 1, delay, error, response))  # type: ignore[arg-type]
        time.sleep(delay)
    raise RuntimeError("retry loop ended without a response")


async def retry_usage_write_async(
    call: Callable[[], Awaitable[Response[T]]], policy: RestRetryPolicy | None = None
) -> Response[T]:
    """Retry a generated ``asyncio_detailed`` usage call."""
    normalized = _normalize(policy or RestRetryPolicy())
    for attempt in range(1, normalized.max_attempts + 1):
        response: Response[T] | None = None
        error: httpx.TransportError | None = None
        try:
            response = await call()
            if not _retryable_status(response.status_code):
                return response
        except httpx.TransportError as caught:
            error = caught
        if attempt >= normalized.max_attempts:
            if error is not None:
                raise error
            return response  # type: ignore[return-value]
        delay = _delay(response, normalized, attempt)
        if normalized.on_retry is not None:
            normalized.on_retry(RestRetryEvent(attempt + 1, delay, error, response))  # type: ignore[arg-type]
        await asyncio.sleep(delay)
    raise RuntimeError("retry loop ended without a response")


def _normalize(policy: RestRetryPolicy) -> RestRetryPolicy:
    initial = max(0.001, policy.initial_backoff_seconds)
    return RestRetryPolicy(
        max_attempts=max(1, policy.max_attempts),
        initial_backoff_seconds=initial,
        max_backoff_seconds=max(initial, policy.max_backoff_seconds),
        jitter=min(1.0, max(0.0, policy.jitter)),
        on_retry=policy.on_retry,
    )


def _retryable_status(status: int) -> bool:
    return status in {429, 500, 502, 503, 504}


def _delay(response: Response[object] | None, policy: RestRetryPolicy, attempt: int) -> float:
    if response is not None:
        retry_after = response.headers.get("Retry-After", "").strip()
        try:
            seconds = float(retry_after)
            if seconds > 0:
                return seconds
        except ValueError:
            try:
                at = parsedate_to_datetime(retry_after)
                now = datetime.now(UTC)
                if at.tzinfo is None:
                    at = at.replace(tzinfo=UTC)
                if at > now:
                    return (at - now).total_seconds()
            except (TypeError, ValueError):
                pass
    backoff = min(policy.max_backoff_seconds, policy.initial_backoff_seconds * (2 ** (attempt - 1)))
    return max(0.001, backoff * (1 - policy.jitter + random.random() * 2 * policy.jitter))
